README.md
# Codeception Redirects Module

Uses the REST module to check URLs for `Location` response headers and `301` HTTP Status response codes, or redirects between protocols.

Handy for checking that `.htaccess` or other methods of web page redirects have been set up correctly.

If test fails stores last shown page in `output` dir.

## Status

* Maintainer: **GaryJones**
* Stability: **alpha**
* Contact: https://gamajo.com

*Please review the code of non-stable modules and provide patches if you have issues.*

## Configuration

None, but ensure that the REST module is enabled and has a `url` set.

### Example (`acceptance.suite.yml`)

~~~yaml
modules:
    enabled:
        - Redirects
        - REST:
            url: 'http://localhost'
~~~

## Example Usage

### 301 Permanent Redirects

Here's a Cest which checks 301 redirects. We turn off automatic following of redirects in the Symfony Browserkit client, so that it doesn't just follow the redirects to the final destination.

```php
<?php

use Codeception\Example;

class RedirectsCest {
    /**
     * @var AcceptanceTester
     */
    protected $I;

    public function _before( AcceptanceTester $I ) {
        $this->I = $I;
    }

    /**
     * @example(old="content/abou", new="about-us")
     * @example(old="content/abou/over.php", new="about-us/company-overview")
     * @example(old="content/abou/miss.php", new="about-us/top-third-mission")
     * @example(old="content/abou/exec.php", new="about-us/executive-team")
     * @example(old="content/abou/team.php", new="about-us/risk-management-specialists")
     *
     * @group redirects
     * @group redirectsabout
     */
    public function redirectOldAboutUrlsToAboutUsPages( AcceptanceTester $I, Example $example ) {
        $this->testIfOldRedirectsToNew($example['old'], $example['new']);
    }

    /**
     * @example(old="content/myac/index.php", new="my-account")
     * @example(old="content/myac/stat.php", new="my-account/account-statements-explained")
     * @example(old="content/myac/depo.php", new="my-account/deposits-withdrawals")
     * @example(old="content/myac/wire.php", new="wire-instructions-r-j-obrien")
     *
     * @group redirects
     * @group redirectsmyaccount
     */
    public function redirectOldMyAccountUrlsToNewMyAccountPages( AcceptanceTester $I, Example $example ) {
        $this->testIfOldRedirectsToNew( $example['old'], $example['new'] );
    }

    private function testIfOldRedirectsToNew($old, $new, $checkDestination = true) {
        $this->I->seePermanentRedirectBetween($old, $new);
        if ($checkDestinationExists) {
            $this->I->urlDoesNotRedirect($new);
        }

        // Check old URL with trailing slash also redirects.
        if (
            '/' !== substr($old, -1) &&
            false === strpos( strrev($old), strrev('.php')) &&
            false === strpos( strrev($old), strrev('.pdf')) &&
            false === strpos( $old, '?')
        ) {
            $old .= '/';
            $this->testIfOldRedirectsToNew($old, $new, $checkDestinationExists);
        }
    }
}

```

Grouping the tests like these mean you can run individual groups of tests more easily:

```sh
vendor/bin/codecept run --env=staging --group=redirectsmyaccount
```

### 307 Temporary Redirects

The above example code also applies to checking for 307 Temporary Redirect, except the `seeTemporaryRedirectBetween()` method is used instead of `seePermanentRedirectBetween()`.

### Protocol Redirects

Here's a Cest which checks protocol redirects, to see if a URL is forced to be served as `http://` or `https://`.

```php
<?php

use Page\ProfileCalendar;
use Page\ProfileContactInformation;
use Page\ProfileMyProducts;
use Step\Acceptance\Login;

class ProtocolRedirectsCest
{
    /**
     * @group protocolRedirects
     */
    public function forceHttp(Login $I)
    {
        $I->wantTo('check forced redirects to HTTP are working.');
        $I->seeHttpProtocolAlwaysUsedFor(ProfileCalendar::$URL);
    }

    /**
     * @group protocolRedirects
     */
    public function forceHttps(Login $I)
    {
        $I->wantTo('check forced redirects to HTTPS are working.');
        $I->seeHttpsProtocolAlwaysUsedFor(ProfileContactInformation::$URL);
        $I->seeHttpsProtocolAlwaysUsedFor(ProfileMyProducts::$URL);
    }
}

```

## Public API

Methods expected to be used in Cepts and Cests.

### followRedirects

Sets whether to automatically follow redirects or not.

```php
$I->followRedirects(false);
```

* `param bool` **`$followRedirects`**

    Whether to follow automatic redirects or not. Default behaviour is true, so most times you'll want to pass in false for 301 redirects tests. Other methods in this package already call this as needed.

### seeRedirectBetween

Check that the specified HTTP Status is returned with the correct Location URL. Fails if either is missing, or `Location` header value does not match the `$url`. Automatically avoids following redirects.

```php
$I->seeRedirectBetween('company/financial-strength-and-security.cfm', 'company/financial-security', 302);
```

* `param` **`$oldUrl`**

    Relative or absolute URL that should be redirected.
* `param` **`$newUrl`**

    Relative or absolute URL of redirect destination.
* `param` **`$statusCode`**

    HTTP status code to check for.

### seePermanentRedirectBetween

A convenience method to check that a 301 HTTP Status is returned with the correct Location URL. Fails if either is missing, or `Location` header value does not match the `$url`. Automatically avoids following redirects.

```php
$I->seePermanentRedirectBetween('company/financial-strength-and-security.cfm', 'company/financial-security');
```

* `param` **`$oldUrl`**

    Relative or absolute URL that should be redirected.
* `param` **`$newUrl`**

    Relative or absolute URL of redirect destination.
    
### seeTemporaryRedirectBetween

A convenience method to check that a 307 HTTP Status is returned with the correct Location URL. Fails if either is missing, or `Location` header value does not match the `$url`. Automatically avoids following redirects.

```php
$I->seeTemporaryRedirectBetween('company/financial-strength-and-security.cfm', 'company/financial-security');
```

* `param` **`$oldUrl`**

    Relative or absolute URL that should be redirected.
* `param` **`$newUrl`**

    Relative or absolute URL of redirect destination.

### urlDoesNotRedirect

 Check that a 200 HTTP Status is returned and the URL has no redirects. Allows the possibility of following redirects.

 ```php
$I->urlDoesNotRedirect('company/financial-security');
 ```

* `param` **`$url`**

    Absolute or relative (to REST config `url`) URL.

### seeHttpProtocolAlwaysUsedFor

 Check that a 200 HTTP Status is eventually returned with the HTTP protocol. Follows redirects automatically.

 ```php
$I->seePermanentRedirectToHttpFor('insecure-page');
 ```

* `param` **`$url`**

    Absolute or relative (to REST config `url`) URL.

### seePermanentRedirectToHttpsFor

 Check that a 200 HTTP Status is eventually returned with the HTTPS protocol. Follows redirects automatically.

 ```php
$I->seePermanentRedirectToHttpsFor('contact-us');
 ```

* `param` **`$url`**

    Absolute or relative (to REST config `url`) URL.


CHANGELOG.md
# Change Log for Codeception Redirects Module

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [0.5.0] - 2020-03-27
### Changed
- Add compatibility with Codeception v3.
- Stop PHP 7.2 from being an allowed failure in Travis CI.

### Fixed
- Change log heading and link.

### Removed
- `composer.lock`, since this is a library, not an application.

## [0.4.1] - 2017-12-21
### Changed
- Reduce minimum PHP requirement from 7.1 to 7.0.
- Tidy-up remnants of tests/ directory.
- Fix whitespace code standards. 

## [0.4.0] - 2017-12-04
### Added
- `.editorconfig`.
- `.gitattributes`.
- Code of conduct.
- Code standards checking and unit test setup.

### Changed
- `seeRedirectBetween()` to `public`. Props [Tim Jensen].
- Minor code tidying.

## [0.3.0] - 2017-10-19
### Added
- `seeTemporaryRedirectBetween()` method.

## [0.2.1] - 2017-05-12
### Added
- Security Advisories package & update composer.

### Fixed
- `followRedirects` for URL exists check.
- Home page URL no redirection check.

## [0.2.0] - 2016-12-01
### Added
- Public `urlDoesNotExist($url)` method.

### Changed
- **Breaking Change** Rename all public methods to be more descriptive.
- Public methods automatically take account of whether they should follow redirects or not.
- Public methods send a HEAD request automatically.

## [0.1.4] - 2016-09-01
- Change the check for URL exists, to just it being a 200 and not having any more redirects.
- Increase minimum version of Codeception to 2.2.0.
- Allow for `'false'` string to represent boolean `false` for check destination exists arg.

## [0.1.3] - 2016-03-25
- Improve `seePermanentRedirectTo()` to check if destination URL exists.
- Add file-level DocBlock.
- Add missing `@since` tag.
- Fix change log links.

## [0.1.2] - 2016-03-21
- Improve README.
- Use `getHeader()` method instead of fetching from an array.
- Use constants for the protocol string.
- Fix the `@since` tags.

## [0.1.1] - 2016-03-19
- Fix invalid composer.json

## 0.1.0 - 2016-03-19
- Initial release.

[Tim Jensen]: https://github.com/timothyjensen

[Unreleased]: https://github.com/gamajo/codeception-redirects/compare/0.5.0...HEAD
[0.5.0]: https://github.com/gamajo/codeception-redirects/compare/0.4.1...0.5.0
[0.4.1]: https://github.com/gamajo/codeception-redirects/compare/0.4.0...0.4.1
[0.4.0]: https://github.com/gamajo/codeception-redirects/compare/0.3.0...0.4.0
[0.3.0]: https://github.com/gamajo/codeception-redirects/compare/0.2.1...0.3.0
[0.2.1]: https://github.com/gamajo/codeception-redirects/compare/0.2.0...0.2.1
[0.2.0]: https://github.com/gamajo/codeception-redirects/compare/0.1.4...0.2.0
[0.1.4]: https://github.com/gamajo/codeception-redirects/compare/0.1.3...0.1.4
[0.1.3]: https://github.com/gamajo/codeception-redirects/compare/0.1.2...0.1.3
[0.1.2]: https://github.com/gamajo/codeception-redirects/compare/0.1.1...0.1.2
[0.1.1]: https://github.com/gamajo/codeception-redirects/compare/0.1.0...0.1.1


src/Redirects.php
<?php
/**
 * Redirects module for Codeception.
 *
 * @package   gamajo\codeception-redirects
 * @author    Gary Jones
 * @copyright 2016 Gamajo
 * @license   MIT
 */

namespace Codeception\Module;

use Codeception\Module;
use Symfony\Component\BrowserKit\Client;
use Symfony\Component\BrowserKit\Response;

/**
 * Redirects module for Codeception.
 *
 * @package Codeception\Module
 * @author  Gary Jones
 */
class Redirects extends Module
{
    /**
     * Protocol (http) constant that can be used with permanentRedirectForProtocol() method.
     */
    const PROTOCOL_HTTP = 'http';

    /**
     * Protocol (https) constant that can be used with permanentRedirectForProtocol() method.
     */
    const PROTOCOL_HTTPS = 'https';

    /**
     * Toggle redirections on and off.
     *
     * By default, BrowserKit will follow redirections, so to check for 30*
     * HTTP status codes and Location headers, they have to be turned off.
     *
     * @since 0.1.0
     *
     * @param bool $followRedirects Optional. Whether to follow redirects or not.
     *                              Default is true.
     */
    public function followRedirects($followRedirects = true)
    {
        $this->getModule('PhpBrowser')->client->followRedirects($followRedirects);
    }

    /**
     * Check if redirections are being followed or not.
     *
     * @since 0.2.0
     *
     * @return bool True if redirects are being followed, false otherwise.
     */
    protected function isFollowingRedirects()
    {
        return $this->getModule('PhpBrowser')->client->isFollowingRedirects();
    }

    /**
     * Check that a redirection occurs.
     *
     * @since 0.2.0
     *
     * @param string  $oldUrl     Relative or absolute URL that should be redirected.
     * @param string  $newUrl     Relative or absolute URL of redirect destination.
     * @param integer $statusCode Status code to check for.
     */
    public function seeRedirectBetween($oldUrl, $newUrl, $statusCode)
    {
        // We must not follow all redirects, so save current situation,
        // force disable follow redirects, and revert at the end.
        $followsRedirects = $this->isFollowingRedirects();
        $this->followRedirects(false);

        $response = $this->sendHeadAndGetResponse($oldUrl);

        if (null !== $response) {
            $responseCode   = $response->getStatus();
            $locationHeader = $response->getHeader('Location', true);

            // Check for correct response code.
            $this->assertEquals($statusCode, $responseCode, 'Response code was not ' . $statusCode . '.');

            // Check location header URL contains submitted URL.
            $this->assertContains($newUrl, $locationHeader, 'Redirect destination not found in Location header.');
        }


        $this->followRedirects($followsRedirects);
    }

    /**
     * Convenience method to check that a 301 HTTP Status is returned with the correct Location URL.
     *
     * @since 0.2.0
     *
     * @param string $oldUrl Relative or absolute URL that should be redirected.
     * @param string $newUrl Relative or absolute URL of redirect destination.
     */
    public function seePermanentRedirectBetween($oldUrl, $newUrl)
    {
        $this->seeRedirectBetween($oldUrl, $newUrl, 301);
    }

    /**
     * Convenience method to check that a 307 HTTP Status is returned with the correct Location URL.
     *
     * @since 0.3.0
     *
     * @param string $oldUrl Relative or absolute URL that should be redirected.
     * @param string $newUrl Relative or absolute URL of redirect destination.
     */
    public function seeTemporaryRedirectBetween($oldUrl, $newUrl)
    {
        $this->seeRedirectBetween($oldUrl, $newUrl, 307);
    }

    /**
     * Check that a 200 HTTP Status is returned and the URL has no redirects.
     *
     * @since 0.1.3
     * @since 0.2.0 Renamed method, made public.
     *
     * @param string $url Relative or absolute URL of redirect destination.
     */
    public function urlDoesNotRedirect($url)
    {
        if ('/' === $url) {
            $url = '';
        }

        // We must not follow all redirects, so save current situation,
        // force disable follow redirects, and revert at the end.
        $followsRedirects = $this->isFollowingRedirects();
        $this->followRedirects(false);

        $response       = $this->sendHeadAndGetResponse($url);

        if (null !== $response) {
            $responseCode   = $response->getStatus();
            $locationHeader = $response->getHeader('Location', true);

            // Check for 200 response code.
            $this->assertEquals(200, $responseCode, 'Response code was not 200.');

            // Check that destination URL does not try to redirect.
            // Somewhat redundant, as this should never appear with a 200 HTTP Status code anyway.
            $this->assertNull($locationHeader, 'Location header was found when it should not exist.');
        }

        $this->followRedirects($followsRedirects);
    }

    /**
     * Use REST Module to send HEAD request and return the response.
     *
     * @since 0.2.0
     *
     * @param string $url
     *
     * @return null|Response
     */
    protected function sendHeadAndGetResponse($url)
    {
        /** @var REST $rest */
        $rest = $this->getModule('REST');
        $rest->sendHEAD($url);

        return $rest->client->getInternalResponse();
    }

    /**
     * Check that a 200 HTTP Status is eventually returned with the HTTP protocol.
     *
     * Should be HTTP. The URL should be the same - only the protocol is different.
     *
     * @since 0.1.0
     * @since 0.2.0 Method renamed.
     *
     * @param string $url Relative or absolute URL of redirect destination.
     */
    public function seeHttpProtocolAlwaysUsedFor($url)
    {
        $this->checkUrlAndProtocolAreCorrect($url, self::PROTOCOL_HTTP);
    }

    /**
     * Check that a 200 HTTP Status is eventually returned with the HTTPS protocol.
     *
     * Should be HTTPS. The URL should be the same - only the protocol is different.
     *
     * @since 0.1.0
     * @since 0.2.0 Method renamed.
     *
     * @param string $url Relative or absolute URL of redirect destination.
     */
    public function seeHttpsProtocolAlwaysUsedFor($url)
    {
        $this->checkUrlAndProtocolAreCorrect($url, self::PROTOCOL_HTTPS);
    }

    /**
     * For a given URL, check that a 200 HTTP Status is returned, the protocol matches given value,
     * and URL (except maybe protocol) has not changed.
     *
     * @since 0.1.0
     * @since 0.2.0 Method renamed.
     *
     * @param string $url      Relative or absolute URL of redirect destination.
     * @param string $protocol Protocol: 'http' or 'https'.
     */
    protected function checkUrlAndProtocolAreCorrect($url, $protocol)
    {
        // We must follow all redirects, so save current situation, force follow redirects, and revert at the end.
        $followsRedirects = $this->isFollowingRedirects();
        $this->followRedirects();

        $url = ltrim($url, '/');

        /** @var REST $rest */
        $rest = $this->getModule('REST');
        $rest->sendHEAD($url);

        /** @var Client $client */
        $client       = $this->getModule('PhpBrowser')->client;
        $responseCode = $client->getInternalResponse()->getStatus();
        $responseUri  = $client->getHistory()->current()->getUri();
        $scheme       = parse_url($responseUri, PHP_URL_SCHEME);

        // Check for 200 response code.
        $this->assertEquals(200, $responseCode, 'Response code was not 200.');

        // Check that destination URL contains submitted URL part.
        $this->assertContains($url, $responseUri, 'Destination URL does not contain the original URL.');

        // Check for submitted http/https value matches destination URL.
        $this->assertEquals($protocol, $scheme, 'Protocol at destination URL does not match expected value.');

        $this->followRedirects($followsRedirects);
    }
}
